dart/flat_buffers.dart

```diff
@@ -0,0 +1,1095 @@
+    _buf.setUint32(_buf.lengthInBytes - tail, x, Endianness.LITTLE_ENDIAN);
+  }
+
+  static void _setInt16AtTail(ByteData _buf, int tail, int x) {
+    _buf.setInt16(_buf.lengthInBytes - tail, x, Endianness.LITTLE_ENDIAN);
+  }
+
+  static void _setUint16AtTail(ByteData _buf, int tail, int x) {
+    _buf.setUint16(_buf.lengthInBytes - tail, x, Endianness.LITTLE_ENDIAN);
+  }
+
+  static void _setInt8AtTail(ByteData _buf, int tail, int x) {
+    _buf.setInt8(_buf.lengthInBytes - tail, x);
+  }
+
+  static void _setUint8AtTail(ByteData _buf, int tail, int x) {
+    _buf.setUint8(_buf.lengthInBytes - tail, x);
+  }
+}
+
+/**
+ * The reader of lists of 64-bit float values.
+ *
+ * The returned unmodifiable lists lazily read values on access.
+ */
+class Float64ListReader extends Reader<List<double>> {
+  const Float64ListReader();
+
+  @override
+  int get size => 4;
+
+  @override
+  List<double> read(BufferContext bc, int offset) =>
+      new _FbFloat64List(bc, bc.derefObject(offset));
+}
+
+class Float32ListReader extends Reader<List<double>> {
+  const Float32ListReader();
+
+  @override
+  int get size => 4;
+
+  @override
+  List<double> read(BufferContext bc, int offset) =>
+      new _FbFloat32List(bc, bc.derefObject(offset));
+}
+
+class Float64Reader extends Reader<double> {
+  const Float64Reader();
+
+  @override 
+  int get size => 8;
+
+  @override
+  double read(BufferContext bc, int offset) => bc._getFloat64(offset);
+}
+
+
+class Float32Reader extends Reader<double> {
+  const Float32Reader();
+
+  @override 
+  int get size => 4;
+
+  @override
+  double read(BufferContext bc, int offset) => bc._getFloat32(offset);
+}
+
+
+/**
+ * The reader of signed 32-bit integers.
+ */
+class Int32Reader extends Reader<int> {
+  const Int32Reader() : super();
+
+  @override
+  int get size => 4;
+
+  @override
+  int read(BufferContext bc, int offset) => bc._getInt32(offset);
+}
+
+/**
+ * The reader of signed 32-bit integers.
+ */
+class Int16Reader extends Reader<int> {
+  const Int16Reader() : super();
+
+  @override
+  int get size => 4;
+
+  @override
+  int read(BufferContext bc, int offset) => bc._getInt16(offset);
+}
+
+/**
+ * The reader of 8-bit signed integers.
+ */
+class Int8Reader extends Reader<int> {
+  const Int8Reader() : super();
+
+  @override
+  int get size => 1;
+
+  @override
+  int read(BufferContext bc, int offset) => bc._getInt8(offset);
+}
+
+/**
+ * The reader of lists of objects.
+ *
+ * The returned unmodifiable lists lazily read objects on access.
+ */
+class ListReader<E> extends Reader<List<E>> {
+  final Reader<E> _elementReader;
+
+  const ListReader(this._elementReader);
+
+  @override
+  int get size => 4;
+
+  @override
+  List<E> read(BufferContext bc, int offset) =>
+      new _FbGenericList<E>(_elementReader, bc, bc.derefObject(offset));
+}
+
+/**
+ * The offset from the end of the buffer to a serialized object of the type [T].
+ */
+class Offset<T> {
```

Comment:
Does Dart guarantee by-value semantics for small wrapper objects like this? In Java for example our offsets are naked ints, to not pay the cost of object allocation everywhere. This sadly of course is less-typesafe.

Reply:
Dart will not treat this as by-val.  Offsets could be made into naked ints, but that seems like a lot of loss in clarity/safety.  Was the Java choice made due to benchmarks demonstrating its perf increases?

Reply:
It was not made based on benchmarks, no. But allocating an object for every offset generated (which there are a lot of) seems excessive.

Reply:
We were told by Dart VM team that allocation is cheap. I can imagine that VM might optimize this to integers. But we have not actually performed benchmarking.

New offset object is created for every object (table), list and string. Is this a lot? Maybe. But probably will not make a lot of performance difference.

Reply:
Allocation is cheap, GC is not. You pay the price elsewhere.

Beyond that, there's the fact that FlatBuffers purpose is to reduce the need for object allocation and copying. So even if Dart's memory management is indeed magically cheaper than other GC languages, it looks weird to have a library like FlatBuffers generate a ton of garbage.

Reply:
I don't have objections against replacing `Offset` with plain `int`.

Reply:
Done.